platform/core/src/services/MeasurementService/MeasurementService.js

```diff
@@ -0,0 +1,137 @@
+import log from '../../log';
+import guid from '../../utils/guid';
+
+const EVENTS = {
+  MEASUREMENT_UPDATED: 'event::measurement_updated',
+  MEASUREMENT_ADDED: 'event::measurement_added',
+};
+
+const CONTEXTS = {
+  ALL: 'context::all',
+  VIEWER: 'context::viewer',
+  CORNERSTONE: 'context::cornerstone',
+};
+
+class MeasurementService {
+  constructor() {
+    this.measurements = {};
+    this.listeners = {};
+    this.events = EVENTS;
+  }
+
+  getEvents() {
+    return { ...this.events };
+  }
+
+  registerEvent(event) {
+    this.events[event] = `event@${event}`;
+  }
+
+  /**
+   * Adds or update persisted measurements.
+   *
+   * @param {Object} measurement
+   * @param {string} measurement.id
+   * @param {MeasurementSchema} measurement.data
+   */
+  addOrUpdate({ id, data }) {
+    if (!this._isValidMeasurement(data)) {
+      log.warn(
+        'Attempting to add or update a null/undefined measurement. Exiting early.'
+      );
+      return;
+    }
+
+    let internalId = id;
+    if (!internalId) {
+      internalId = guid();
+      log.warn(`Measurement ID not set. Using random string ID: ${internalId}`);
+    }
+
+    const measurement = {
+      id,
+      modifiedTimestamp: Math.floor(Date.now() / 1000),
+      data,
+    };
+
+    if (this.measurements[internalId]) {
+      log.warn(`Measurement already defined. Updating measurement.`);
+      this.measurements[internalId] = measurement;
+      this._broadcastChange(internalId, EVENTS.MEASUREMENT_UPDATED);
+    } else {
+      log.warn(`Measurement added.`);
+      this.measurements[internalId] = measurement;
+      this._broadcastChange(internalId, EVENTS.MEASUREMENT_ADDED);
+    }
+
+    return measurement;
```

Comment:
Probably okay to return _just_ the `measurementId` here. The `addOrUpdate` caller should have everything else.